File: monadic-counter/src/Counter.hs
-- | Monadic counters.

module Counter where

import           Control.Monad.Reader
import           Control.Monad.State.Strict
import           Control.Monad.Writer.Strict

newtype Counter = MkCounter {cValue :: Int}
                deriving (Show)

-- | 'inc c n' increments the counter by 'n' units.
inc :: Counter -> Int -> Counter
inc (MkCounter c) n = MkCounter (c + n)

-- | The computation we want to run.
mComputation :: Counter
mComputation = inc (inc (inc (inc (inc (MkCounter 0) 3) 3) 3) 5) 5

-- | Run and show the computation.
runComputation :: IO ()
runComputation = putStrLn $ show mComputation

-- * The monadic counter.

-- | CounterS is a monad.
type CounterS = State Counter

-- | Increment the counter by 'n' units.
incS :: Int-> CounterS ()
incS n = modify (\c -> inc c n)

-- | The computation we want to run, with the state monad.
mComputationS :: CounterS ()
mComputationS = do
  incS 3
  incS 3
  incS 3
  incS 5
  incS 5

runComputationS :: IO ()
runComputationS = putStrLn $ show . snd $ runState mComputationS (MkCounter 0)

-- * Introducing an environment.
type CounterRS = ReaderT Int CounterS

-- | Increment the counter by the amount of units specified by the environment.
incR :: CounterRS ()
incR = ask >>= lift . incS

-- | The computation we want to run, using reader and state monads.
mComputationRS :: CounterRS ()
mComputationRS = do
  local (const 3) $ do
    incR
    incR
    incR
    local (const 5) $ do
      incR
      incR

runComputationRS :: IO ()
runComputationRS = putStrLn $ show . snd $ runState compWithEnv (MkCounter 0)
  where compWithEnv = runReaderT mComputationRS 15

-- * Adding logging.
type CounterWRS = WriterT [Int] CounterRS

incW :: CounterWRS ()
incW = lift incR >> get >>= tell . (:[]) . cValue

mComputationWRS :: CounterWRS ()
mComputationWRS = do
  local (const 3) $ do
    incW
    incW
    incW
    local (const 5) $ do
      incW
      incW

runComputationWRS :: IO ()
runComputationWRS = putStrLn $ show $ runState compWithEnv (MkCounter 0)
  where compWithEnv = runReaderT compWithLogs 15
        compWithLogs = (runWriterT mComputationWRS)

File: monadic-counter-scala-remix/src/main/scala/com/example/Counter.scala
package com.example

import cats.data._
import cats.data.WriterT
import cats.instances.all._
import cats._
import cats.arrow.Arrow

/**
  * Created by damian on 31-8-16.
  */
object Counter {

  case class Counter(cValue: Int) extends AnyVal {
    def inc(n: Int): Counter = Counter(cValue + n)
  }

  // The computation we want to run.
  val mComputation: Counter =
  Counter(0)
    .inc(3)
    .inc(3)
    .inc(3)
    .inc(5)
    .inc(5)

  // Introducing state.
  type CounterS[A] = State[Counter, A]

  def incS(n: Int): CounterS[Unit] = {//State[Counter, Unit] = {
    State.modify(c => c.inc(n))
  }

  val mComputationS: CounterS[Unit] = for {
    _ <- incS(3)
    _ <- incS(3)
    _ <- incS(3)
    _ <- incS(5)
    _ <- incS(5)
  } yield ()

  // Introducing an environment.
  type CounterRS[A] = ReaderT[CounterS[?], Int, A]

  def incR: CounterRS[Unit] = {
    Kleisli
      .ask[CounterS[?], Int]
      .flatMap{n:Int => ReaderT.lift(incS(n))}
  }

  def local(n: Int) = Kleisli.local[CounterS[?], Unit, Int](_ => n)(_)

  val mComputationRS: CounterRS[Unit] =
    local(3) {
      for {
        _ <- incR
        _ <- incR
        _ <- incR
        _ <- local(5) {
          for {
            _ <- incR
            _ <- incR
          } yield ()
        }
      } yield ()
    }

  // Adding logging.
  type CounterWRS[A] = WriterT[CounterRS[?], Seq[Int], A]

  // WriterT does not have lift in the latest release of Cats.
  def incW: CounterWRS[Unit] = ???

}
